## **Mastering the Pinecone vector database.**

## **GOAL: save text sentence embeddings and perform semantic search in Pinecone using Python code.**

Create a Pinecone Account: Head to 'Pinecone.io' and sign up for a free account. After signing up, you’ll be given an API key.

 To start working with Pinecone,
install 'pinecone' library

In [7]:
!pip install pinecone python-dotenv
print("Pinecone library installed successfully!")

Pinecone library installed successfully!



[notice] A new release of pip is available: 23.0.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


From 'pinecone' library import Pinecone, ServerlessSpec.

In [3]:
from pinecone import Pinecone, ServerlessSpec
print("Pinecone library imported successfully!")

Pinecone library imported successfully!


On the Pinecone main page, navigate to the "Get started" section via the left menu. Next, follow the "Quick Start for Developers" guide.
* Initialaze a client.
* Create an index.
* Upsert data.
* Perform a search.









When you perform "Upsert data", add a list of your own text strings.

When you perform "Search", use of your own text string.

In [ ]:
from dotenv import load_dotenv
import os
load_dotenv()

# initialize client
pc = Pinecone(api_key=os.getenv("PINECONE_API"))

index_name = "accenture-bootcamp-py"

if not pc.has_index(index_name):
    pc.create_index_for_model(
        name=index_name,
        cloud="aws",
        region="us-east-1",
        embed={
            "model":"llama-text-embed-v2",
            "field_map":{"text": "chunk_text"}
        }
    )

In [17]:
index = pc.Index(index_name)

# Upsert text data
text_data = [
    {"_id": "rec1", "chunk_text": "Machine learning is a subset of artificial intelligence that enables systems to learn and improve from experience."},
    {"_id": "rec2", "chunk_text": "Deep learning uses neural networks with multiple layers to process data and extract features."},
    {"_id": "rec3", "chunk_text": "Natural language processing is a field of AI focused on understanding and generating human language."},
    {"_id": "rec4", "chunk_text": "Vector databases store data as high-dimensional vectors, enabling efficient similarity search."},
    {"_id": "rec5", "chunk_text": "Semantic search finds documents based on meaning rather than keyword matching."}
]

# Upsert the data into the index
index.upsert_records("__default__",text_data)
print(f"Successfully upserted {len(text_data)} text documents to the index.")

Successfully upserted 5 text documents to the index.


Print search results.

In [40]:
# search
query = "How do neural networks work in deep learning?"

results = index.search(
    "__default__",
    query={
        "top_k": 3,
        "inputs": {
            "text": query,
        }
    }
)

print(f"Query: '{query}'\n")
print("| id    | score | text                                                                    |")
print("|-------|-------|","-"*73,"|")
for hit in results['result']['hits']:
    print(f"| {hit['_id']:<5} | {round(hit['_score'], 2):<5} | {hit['fields']['chunk_text']:<50} |")

Query: 'How do neural networks work in deep learning?'

| id    | score | text                                                                    |
|-------|-------| ------------------------------------------------------------------------- |
| rec2  | 0.48  | Deep learning uses neural networks with multiple layers to process data and extract features. |
| rec3  | 0.23  | Natural language processing is a field of AI focused on understanding and generating human language. |
| rec1  | 0.17  | Machine learning is a subset of artificial intelligence that enables systems to learn and improve from experience. |


## Result

| id    | score | text                                                                    |
|-------|-------|-------------------------------------------------------------------------|
| rec2  | 0.48  | Deep learning uses neural networks with multiple layers to process data and extract features. |
| rec3  | 0.23  | Natural language processing is a field of AI focused on understanding and generating human language. |
| rec1  | 0.17  | Machine learning is a subset of artificial intelligence that enables systems to learn and improve from experience. |